translate/translate.go

```diff
@@ -79,42 +85,66 @@ func couldBeValidTranslator(t reflect.Type) bool {
 	if t.Kind() != reflect.Func {
 		return false
 	}
-	if t.NumIn() != 1 || t.NumOut() != 1 {
+	if t.NumIn() != 1 || t.NumOut() != 2 {
 		return false
 	}
-	if util.IsInvalidInConfig(t.In(0).Kind()) || util.IsInvalidInConfig(t.Out(0).Kind()) {
+	if util.IsInvalidInConfig(t.In(0).Kind()) ||
+		util.IsInvalidInConfig(t.Out(0).Kind()) ||
+		t.Out(1) != translationsType {
 		return false
 	}
 	return true
 }
 
+// fieldName returns the name uses when (un)marshalling a field. t should be a reflect.Value of a struct,
+// index is the field index, and tag is the struct tag used when (un)marshalling (e.g. "json" or "yaml")
+func fieldName(t reflect.Value, index int, tag string) string {
+	f := t.Type().Field(index)
+	if tag == "" {
+		return f.Name
+	}
+	return strings.Split(f.Tag.Get(tag), ",")[0]
+}
+
 // translate from one type to another, but deep copy all data
 // precondition: vFrom and vTo are the same type as defined by translatable()
 // precondition: vTo is addressable and settable
-func (t translator) translateSameType(vFrom, vTo reflect.Value) {
+func (t translator) translateSameType(vFrom, vTo reflect.Value, fromPath, toPath path.ContextPath) {
 	k := vFrom.Kind()
 	switch {
 	case util.IsPrimitive(k):
 		// Use convert, even if not needed; type alias to primitives are not
 		// directly assignable and calling Convert on primitives does no harm
 		vTo.Set(vFrom.Convert(vTo.Type()))
+		t.translations.AddTranslation(fromPath, toPath)
 	case k == reflect.Ptr:
 		if vFrom.IsNil() {
 			return
 		}
 		vTo.Set(reflect.New(vTo.Type().Elem()))
-		t.translate(vFrom.Elem(), vTo.Elem())
+		t.translate(vFrom.Elem(), vTo.Elem(), fromPath, toPath)
 	case k == reflect.Slice:
 		if vFrom.IsNil() {
 			return
 		}
 		vTo.Set(reflect.MakeSlice(vTo.Type(), vFrom.Len(), vFrom.Len()))
 		for i := 0; i < vFrom.Len(); i++ {
-			t.translate(vFrom.Index(i), vTo.Index(i))
+			t.translate(vFrom.Index(i), vTo.Index(i), fromPath.Append(i), toPath.Append(i))
 		}
 	case k == reflect.Struct:
 		for i := 0; i < vFrom.NumField(); i++ {
-			t.translate(vFrom.Field(i), vTo.FieldByName(vFrom.Type().Field(i).Name))
+			fieldGoName := vFrom.Type().Field(i).Name
+			toStructField, _ := vTo.Type().FieldByName(fieldGoName)
```

Comment:
Might be better to not throw away the `bool` return of `FieldByName` and `panic` with a more reasonable message than allowing the follow-up line to panic at index out of bounds / null object.

Reply:
Fair, it should never be hit since the translatable check should handle that, but I'm all for better panic messages.